File: src/09_Range_Queries/sum_queries_2d.cpp
#include <bits/stdc++.h>
#include "../compete.hpp"
using namespace std;


const int n = 5;
vector<vector<int> > v(n, vector<int>(n));
// prefix sum matrix
vector<vector<int> > psm(n, vector<int>(n));
// bottom-right & top-left coordinates
int x_br, y_br, x_tl, y_tl;


template <int x=10>
int rand_int() {return rand() % x;}


// O(n^2) (linear with the input size)
void populate_psm()
{
  // populate the first row
  partial_sum(v[0].begin(), v[0].end(), psm[0].begin());
  // populate the first column
  for (int i = 1; i < n; ++i)
    psm[i][0] = psm[i - 1][0] + v[i][0];
  // populate the rest
  for (int i = 1; i < n; ++i)
    for (int j = 1; j < n; ++j)
      psm[i][j] = psm[i - 1][j] + psm[i][j - 1] - psm[i - 1][j - 1] + v[i][j];
}


// O(1)
void solve()
{
  int a = psm[y_br][x_br];
  if (x_tl) a -= psm[y_br][x_tl - 1];
  if (y_tl) a -= psm[y_tl - 1][x_br];
  if (x_tl && y_tl) a += psm[y_tl - 1][x_tl - 1];
  cout << "Submatrix sum: " << a << '\n';
}


int main()
{
  // Populate the matrix with random integers in [0-10)
  for (vector<int>& ve : v) generate(ve.begin(), ve.end(), rand_int<10>);
  // Populate the prefix sum matrix
  populate_psm();

  cout << v << '\n';

  cout << "Insert bot right coords (x y): ";
  cin >> x_br >> y_br;
  cout << "Insert top left coords (x y): ";
  cin >> x_tl >> y_tl;

  solve();
}

File: src/09_Range_Queries/sum_queries_2d.py
from random import randint
from pprint import pprint

n = 5  # rows
m = 7  # cols
v = [[randint(0, 9)] * m for _ in range(n)]
psm = [[0] * m for _ in range(n)]


# O(nm)
def populate_psm():
  psm[0][0] = v[0][0]

  # 1st row
  for j in range(1, m):
    psm[0][j] = v[0][j] + psm[0][j - 1]

  # 1st col
  for i in range(1, n):
    psm[i][0] = v[i][0] + psm[i - 1][0]

  # rest
  for i in range(1, n):
    for j in range(1, m):
      psm[i][j] += psm[i - 1][j] + psm[i][j - 1] - psm[i - 1][j - 1] + v[i][j]


# O(1)
def solve(x_br, y_br, x_tl, y_tl):
  ans = psm[y_br][x_br]
  if x_tl:
    ans -= psm[y_br][x_tl - 1]
  if y_tl:
    ans -= psm[y_tl - 1][x_br]
  if x_tl and y_tl:
    ans += psm[y_tl - 1][x_tl - 1]
  return ans


if __name__ == '__main__':
  populate_psm()

  pprint(v)
  print()
  pprint(psm)
  print()
  assert psm[-1][-1] == sum(sum(b) for b in v)

  # (x_br, y_br), (x_tl, y_tl)
  queries = (
    ((5, 4), (1, 2)),
    ((4, 3), (2, 2)),
    ((0, 4), (0, 2)),
    ((0, 4), (0, 0)),
    ((4, 0), (1, 0)),
    ((4, 3), (1, 0)),
  )

  for br, tl in queries:
    print(br, tl, ':', solve(*br, *tl))
